src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Tasks/DoNotUseWhenAllOrWaitAllWithSingleArgumentFixer.cs

```diff
@@ -0,0 +1,102 @@
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.NetCore.Analyzers.Tasks
+{
+    public abstract class DoNotUseWhenAllOrWaitAllWithSingleArgumentFixer : CodeFixProvider
+    {
+        public override ImmutableArray<string> FixableDiagnosticIds => ImmutableArray.Create(
+            DoNotUseWhenAllOrWaitAllWithSingleArgument.WaitAllRule.Id,
+            DoNotUseWhenAllOrWaitAllWithSingleArgument.WhenAllRule.Id);
+
+        public override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            var cancellationToken = context.CancellationToken;
+            var root = await context.Document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var nodeToFix = root.FindNode(context.Span, getInnermostNodeForTie: true);
+
+            var semanticModel = await context.Document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var operation = (IInvocationOperation)semanticModel.GetOperation(nodeToFix, cancellationToken);
+            if (operation.TargetMethod.Name == nameof(Task.WaitAll))
+            {
+                var title = MicrosoftNetCoreAnalyzersResources.DoNotUseWaitAllWithSingleTaskFix;
+                context.RegisterCodeFix(new MyCodeAction(title,
+                    async ct =>
+                    {
+                        var editor = await DocumentEditor.CreateAsync(context.Document, ct).ConfigureAwait(false);
+                        editor.ReplaceNode(nodeToFix,
+                            editor.Generator.InvocationExpression(
+                                editor.Generator.MemberAccessExpression(
+                                    GetSingleArgumentSyntax(operation),
+                                    nameof(Task.Wait))
+                                ).WithTriviaFrom(nodeToFix)
+                            );
+
+                        return editor.GetChangedDocument();
+                    },
+                    equivalenceKey: title),
+                context.Diagnostics);
+            }
+            else if (!IsValueStored(operation))
+            {
+                var title = MicrosoftNetCoreAnalyzersResources.DoNotUseWhenAllWithSingleTaskFix;
+                context.RegisterCodeFix(new MyCodeAction(title,
+                    async ct =>
+                    {
+                        var editor = await DocumentEditor.CreateAsync(context.Document, ct).ConfigureAwait(false);
+                        var newNode = GetSingleArgumentSyntax(operation).WithTriviaFrom(nodeToFix);
+
+                        // The original invocation already returns a Task,
+                        // so we can just replace directly with the argument
+                        editor.ReplaceNode(nodeToFix, newNode);
+
+                        return editor.GetChangedDocument();
+                    },
+                    equivalenceKey: title),
```

Comment:
IIRC, there was a comment from @sharwell somewhere that equivalence key shouldn't be localized and use nameof(Resources.FixTitle) instead. I don't know the reason though, but just pointing this out.

Reply:
https://github.com/dotnet/roslyn/pull/52077#discussion_r606047914